### PR TITLE
[issue-1248] features inference for passed into check dataframes

### LINKED ISSUE #1248: [FEAT] auto infer features when dataframe is passed to check + warning
**Is your feature request related to a problem? Please describe.**
Currently when a df is passed to a check all of its columns are treated as non features, this behaviour is different than the behaviour when creating a "plain" dataset from a data frame.

**Describe the solution you'd like**
when a df is passed to a check will create a dataset from it in the same default manner as when creating a "plain" dataset. As discussed this behavior should include auto infer of features.
In addition should add a warning when a df is passed, explaining that categorical features will be inferred and explaining / linking to how to define a dataset.

### PR DESCRIPTION
**Pull request by yromanyshyn**
resolves #1248 

@Nadav-Barak 
it is enough to modify line 889 in `Dataset.cast_to_dataset` to solve that issue, but I am not sure how clear the warning message raised by the features inference logic (`Dataset._infer_categorical_features` line 573) will be for the users in this case. What do you think?

### CODE REVIEW

**Comment by matanper** [comment]
I think that in Dataset line 889 we can add the warning "DataFrame was passed, initializing deepchecks.tabular.Dataset using the given dataframe"

**Comment by Nadav-Barak** [comment]
I think the warning massage is good for both cases (if the user passed a DF directly or created a DS without specifying categorical)

File: tests/base/dataset_test.py [hunk 1]
```diff
@@ -842,7 +842,7 @@ def test__ensure_not_empty_dataset__with_dataframe(iris: pd.DataFrame):
     ds = Dataset.cast_to_dataset(iris)
     # Assert
     assert_that(ds, instance_of(Dataset))
-    assert_that(ds.features, has_length(0))
+    # assert_that(ds.features, has_length(0))
```
Comment (Nadav-Barak) [inline tests/base/dataset_test.py:845]: maybe change to assert that num features is greater then 0

File: tests/checks/overview/columns_info_test.py [hunk 2]
```diff
@@ -48,9 +48,17 @@ def test_columns_info():
...
+        'date': 'numerical feature', 
+        'a': 'categorical feature', 
+        'b': 'numerical feature', 
+        'c': 'numerical feature', 
+        'label': 'categorical feature'
+    }
     # in df all columns are other
-    expected_res_df = {'index': 'other', 'date': 'other', 'a': 'other',
-                       'b': 'other', 'c': 'other', 'label': 'other'}
+    # expected_res_df = {'index': 'other', 'date': 'other', 'a': 'other',
```
Comment (Nadav-Barak) [inline tests/checks/overview/columns_info_test.py:60]: Think this commented section can be deleted

**Review by Nadav-Barak — APPROVED** [review]

File: tests/checks/overview/columns_info_test.py (hunk 2 above)
Comment (ItayGabbay) [reply tests/checks/overview/columns_info_test.py:60]: Agree

File: tests/checks/integrity/string_mismatch_comparison_test.py [hunk 3]
```diff
@@ -82,16 +82,16 @@ def test_no_mismatch_on_numeric_string_column():
...
 
-    # Act
-    result = StringMismatchComparison().run(pd.DataFrame(data=data), pd.DataFrame(data=compared_data)).value
+#     # Act
+#     result = StringMismatchComparison().run(pd.DataFrame(data=data), pd.DataFrame(data=compared_data)).value
 
-    # Assert
-    assert_that(result, has_length(0))
+#     # Assert
+#     assert_that(result, has_length(0))
```
Comment (ItayGabbay) [inline tests/checks/integrity/string_mismatch_comparison_test.py:95]: Let's delete commented code

**Review by ItayGabbay — APPROVED** [review]

File: tests/checks/integrity/string_mismatch_comparison_test.py (hunk 3 above)
Comment (yromanyshyn) [reply tests/checks/integrity/string_mismatch_comparison_test.py]: removed

File: tests/checks/overview/columns_info_test.py (hunk 2 above)
Comment (yromanyshyn) [reply tests/checks/overview/columns_info_test.py]: removed

File: tests/base/dataset_test.py (hunk 1 above)
Comment (yromanyshyn) [reply tests/base/dataset_test.py]: fixed

**Comment by yromanyshyn** [comment]
@matanper 

> I think that in Dataset line 889 we can add the warning "DataFrame was passed, initializing deepchecks.tabular.Dataset using the given dataframe"

Think it will be redundant, this function is part of a public API and it could be strange for user to see warning as this when he/she will try to use this function directly

**Comment by matanper** [comment]
> @matanper
> 
> > I think that in Dataset line 889 we can add the warning "DataFrame was passed, initializing deepchecks.tabular.Dataset using the given dataframe"
> 
> Think it will be redundant, this function is part of a public API and it could be strange for user to see warning as this when he/she will try to use this function directly

I don't see a reason for it to be a public API, it doesn't help the user more than calling the ctor

**Comment by yromanyshyn** [comment]
@matanper 

> I don't see a reason for it to be a public API, it doesn't help the user more than calling the ctor

ok, I have not changed its visibility marker, but I have added a warning

**Pull request merged**